Fix FileSystem dock won't show any file folders while loading large projects (v2)

Hilderin opened the pull request:
Fixes #92335, Fixes #47053 and Fixes #85082
This is a new PR for the reverted PR #92650.

This PR improves:
- FileSystem Dock not showing progress bar on first file scanning
- Better progression (%) for scanning large projects with a lot of folders and files
- Performance optimization on scanning project files for large projects
- Add progression popup for _update_scan_actions step (instead of Editor freeze)
- Add progression popup for _update_script_classes step (instead of Editor freeze)
- Add progression popup for loading editor layout and reopening scenes (instead of Editor freeze)

I tested these modifications with a project having all the files from https://github.com/PMDCollab/SpriteCollab/tree/master/sprite (more than 125000 files and 385Mb).

Before:
FileSystem was empty without any feedback for a couple of minutes and after that the editor was just frozen:

![image](https://github.com/godotengine/godot/assets/81109165/75c7a53c-cf89-4df5-ad09-c94382f423a9)


After modifications:
![image](https://github.com/godotengine/godot/assets/81109165/58b70ca3-0b2a-4ea5-86a2-346b29bc1b60)
![image](https://github.com/godotengine/godot/assets/81109165/2d4c76ae-eccf-43ac-8b69-d3ac4b8ad633)


**Performance optimization on project scanning before asset importation**
For a 30000 files project, it took around 2m50s on startup to scan the project.
After a couple of optimizations, it now takes 1m30s for the same project.

Before:
![image](https://github.com/godotengine/godot/assets/81109165/4442d2d2-0fc7-4c46-b42f-51648e9e40e3)

After:
![image](https://github.com/godotengine/godot/assets/81109165/e07dbc15-b59a-4f04-9c8d-e249f9786f53)


Causes:
- The FileAccess:exists was calling the open method which has a couple of overhead over file_exists.
- The _scan_new_dir was double sorting the folders.
- The reimport_files was recalculating the import_groupe_file even if _update_scan_actions did that just before.

Hilderin committed:
Fix FileSystem dock won't show any file folders (v2)

RandomShaper commented on editor/editor_file_system.cpp:
```diff
@@ -749,28 +769,50 @@ void EditorFileSystem::scan() {
+		}
+	}
+
+	float old_estimated = estimated;
+	estimated = real + nb_file_to_scan + (nb_dir_to_scan * average * p_depth);
+	if (estimated < old_estimated * 0.95) {
+		estimated = old_estimated * 0.95;
+	}
+	if (estimated <= real + average) {
+		estimated = real + average;
+	}
```
```suggestion
	float min_estimated = MAX(old_estimated * 0.95f, real + average);
	float new_estimated = real + nb_file_to_scan + (nb_dir_to_scan * average * p_depth);
	estimated = MAX(min_estimated, new_estimated);
```

Please double-check the logic is equivalent, though.

Hilderin replied:
Yes, it seems to be equivalent. Done!

RandomShaper commented on editor/editor_file_system.cpp:
```diff
@@ -749,28 +769,50 @@ void EditorFileSystem::scan() {
+void EditorFileSystem::ScanProgress::increment() {
+	real++;
+	nb_file_to_scan--;
+	current = CLAMP(real / estimated, current, 1);
```
```suggestion
	current = CLAMP(real / estimated, current, 1.0f);
```

RandomShaper commented on editor/editor_file_system.cpp:
```diff
@@ -749,28 +769,50 @@ void EditorFileSystem::scan() {
+	// The algorithm calculates an estimated value for the progress of scanning directories and files.
+	// It adjusts the estimate based on the number of directories and files discovered so far,
+	// and ensures the estimated value does not decrease drastically between updates.
+	float average = 10;
```
```suggestion
```

RandomShaper commented on editor/editor_file_system.cpp:
```diff
@@ -749,28 +769,50 @@ void EditorFileSystem::scan() {
+	nb_dir_to_scan += p_nb_sub_dirs;
+	nb_file_to_scan += p_nb_files;
+
+	if (directories_processed > 2) {
+		average = (real + nb_file_to_scan) / directories_processed;
+		if (average < 10) {
+			average = 10;
+		}
```
```suggestion
	float average = 10.0f;
	if (directories_processed > 2) {
		average = MAX(average, (real + nb_file_to_scan) / directories_processed);
```

Chaosus commented on editor/editor_node.cpp:
```diff
@@ -5124,11 +5126,19 @@ void EditorNode::_load_editor_layout() {
 		return;
 	}
 
+	ep.step("Loading docks...", 1, true);
 	editor_dock_manager->load_docks_from_config(config, "docks");
+
+	ep.step("Reopening scenes...", 2, true);
 	_load_open_scenes_from_config(config);
+
+	ep.step("Loading central editor layout...", 3, true);
 	_load_central_editor_layout_from_config(config);
 
+	ep.step("Loading plugin window layout...", 4, true);
 	editor_data.set_plugin_window_layout(config);
+
+	ep.step("Editor layout ready.", 5, true);
```
```suggestion
	ep.step(TTR("Loading docks..."), 1, true);
	editor_dock_manager->load_docks_from_config(config, "docks");

	ep.step(TTR("Reopening scenes..."), 2, true);
	_load_open_scenes_from_config(config);

	ep.step(TTR("Loading central editor layout..."), 3, true);
	_load_central_editor_layout_from_config(config);

	ep.step(TTR("Loading plugin window layout..."), 4, true);
	editor_data.set_plugin_window_layout(config);

	ep.step(TTR("Editor layout ready."), 5, true);
```

Hilderin commented:
All the correction should be done. Thanks for the review.

Chaosus commented on editor/editor_node.cpp:
```diff
@@ -5103,6 +5103,8 @@ void EditorNode::save_editor_layout_delayed() {
 
 void EditorNode::_load_editor_layout() {
+	EditorProgress ep("loading_editor_layout", TTR("Loading editor"), 5);
+	ep.step("Loading editor layout...", 0, true);
```
```suggestion
	ep.step(TTR("Loading editor layout..."), 0, true);
```

Hilderin replied:
done!

Hilderin commented:
I think it's important to note that some issues that this PR fixes have also been addressed by #92303, but in a different way. I suggest merging #92303 before this one, and then I will rebase and adjust this PR.

Hilderin commented:
I rebased from master and reworked this PR.

With the modifications from #92303, I was able to simply this PR a lot. It's now almost only adding `EditorProgress` at different places during first startup and importation process.

I also kept some optimizations like the `FileAccess::exists`, `ResourceFormatImporter::get_resource_import_info` and `ResourceLoader::_path_remap`.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b) on various official demos, it works as expected. The `--import` command line argument still works correctly too.

Code looks good to me.

Hilderin commented:
Rebased to fix conflict with master.

akien-mga commented on core/io/file_access.cpp:
```diff
@@ -59,11 +59,9 @@ bool FileAccess::exists(const String &p_name) {
-		return false;
-	}
-	return true;
+	// Using file_exists because it's faster then trying to open the file.
```
```suggestion
	// Using file_exists because it's faster than trying to open the file.
```

Hilderin replied:
Done!

akien-mga commented on core/io/resource_loader.cpp:
```diff
@@ -1060,36 +1060,39 @@ String ResourceLoader::_path_remap(const String &p_path, bool *r_translation_rem
-					ERR_PRINT("Parse error: " + p_path + ".remap:" + itos(lines) + " error: " + error_text + ".");
-					break;
-				}
+		// Usually, there's no remap file and FileAccess::exists() is faster then FileAccess::open().
```
```suggestion
		// Usually, there's no remap file and FileAccess::exists() is faster than FileAccess::open().
```

Hilderin replied:
Done!

akien-mga commented:
I'll try to test it a bit next week, but this should be ready for a merge soon.

Saul2022 commented:
Testing on  https://github.com/PMDCollab/SpriteCollab/tree/master now without the pr that heavily improves initial anlysis  before scanning files dialog, because of curiosity to see what's going on. 

Initial transition from godot logo to editor scan( over 15 minutes( luckily with the load project pr is down to 1:30 or 3 minutes, so i thank that.).

Then on godot scan, update action and preparing for import dialog 11:30 minutes. Though on preparing for import it gets stuck at 99% which seems to still freeze with no additional progress bar, and atm it still going now with over 12  minutes and still going 

Aside from this something i thought could be done in the godot dialog, which i would guess it might freeze too after the circle wheel dissapears , is make it have a progress bar that says what it doing, because it pretty weird how the editor gets staring  without seeing  anything going on, which might make people thing it got frozen..

Hilderin commented:
Wow, thank you to have tested it, I found why it's stuck before the reimportation. All the resources must pass in `EditorNode::_resources_reimporting` and `EditorNode::_resources_reimported` and for each one `ResourceLoader::get_resource_type(res_path)` is called forcing to open the asset file a couple of times each. I did another modification to optimize this process by using the resource type already calculated in `EditorFileSystem` and I added another progress step while doing in.

Time on reimporting and reimported (each):
Before: 90sec
After: 8sec

Hilderin commented:
> Aside from this something i thought could be done in the godot dialog, which i would guess it might freeze too after the circle wheel dissapears , is make it have a progress bar that says what it doing, because it pretty weird how the editor gets staring without seeing anything going on, which might make people think it got frozen.

Are you speaking about the Godot splash screen at startup? If I understand correctly you suggest to add a progress bar here?
![image](https://github.com/user-attachments/assets/edd861ca-f934-498a-8efe-d053e823d0f8)

Saul2022 commented:
>  If I understand correctly you suggest to add a progress bar here?

 Yea, as there's no visual indication to the user , of whats going on , like staying 1:30 minutes staring at the godot logo, without knowing what's going on , might make people think it got frozen.

Hilderin commented:
Thanks again for all your benchmarks, very appreciated.

> Yea, as there's no visual indication to the user , of whats going on , like staying 1:30 minutes staring at the godot logo, without knowing what's going on , might make people think it got frozen.

It's challenging technically to add a progress bar on the logo splash screen, but, I added another progress popup while executing the first scan. That way, the splash screen will disappear sooner when the first scan starts and as a bonus, I added steps in the progress bar for the creation of autoload scripts and plugins which could take a while if you have a lot of autoloads or plugins.

> though in general what would be the general speed improvement on all the asset reimport/reimported after the preparing import dialog is done ( like what you tested with the sprites of your linked project)?

This optimization should affect all asset types. When reimporting a small number of assets, the difference will be minimal, but in large quantity, removing the need to read the asset should improve performance in the same magnitude for any kind of asset. Still better, I just pushed another little modification and I was able to cut the processing time in `EditorNode::_resources_reimported` just by caching the resource types calculated in `EditorNode::_resources_reimporting`.

> Atp what is just left to do would be fix the cache thing as Juan suggested, and maybe cache the scenes that were opened before closing., as they only seem to be cached upon switching scenes, so when opening they load way faster.

I'm working on it.

Calinou commented:
> It's challenging technically to add a progress bar on the logo splash screen

We could display a loading percentage in the window title, but it'll be missed by those having the editor in fullscreen while the editor is loading (which is unlikely).

Hilderin commented:
> We could display a loading percentage in the window title, but it'll be missed by those having the editor in fullscreen while the editor is loading (which is unlikely).

That could be a brilliant idea!! Right now, I added a `EditorProgress` in `_first_scan_filesystem` which should do the trick. Also, on Android the window title is not displayed :(

Saul2022 commented:
>. Also, on Android the window title is not displayed :(

That  is weird  , if you mean your project like how it appears at the bottom left on windows, ut normal for it to not appear on android.

Else if you mean godot engine logo it there 
![Screenshot_20240821_040002](https://github.com/user-attachments/assets/583a6670-2fb5-4bf7-86ae-5e153fc8d3cc)

Either way this changes are really appreciated , so thank you for this amazing job.

Hilderin commented:
I think Calinou was referring to the window title bar that we have on desktop:
![image](https://github.com/user-attachments/assets/2040fc49-9dd2-485e-a266-033829bb1b80)

Saul2022 commented:
Final update on it  17: mins for initial scan , 6 minutes for scan, 7 for action scans, then 4 on preimport preratation , 6  for exectuting reimport precess and 2 hours for reimport to get to 50%  using  sprite collab whole project.  
 
Some notes( though this is for a future pr, as this  one is pretty well as it is, and is not a regression of 4.3) : i think this is now pretty much  max speed this can get though maybe for some stuff like action scan  or reimport could use some improvements like as the pngs reimport  is instant, i think  what is limiting here is that it going over 1 by 1    like you reimported this image now? Ok then go for the next one ,add up to that , that this files are just some kb ( not even getting to 1 mbs )..and the burden here being that there are tons and tons of them and while the import of each is instant, having so many  and just importing one at a time instead of say 10 and then  pass over to other ten and go on until import is over. It pretty confusing like this so i might create a issue later, though this  might pretty much affect models too., or i might get too crazy and is just that  multi threading  bug or lack of it for import.

akien-mga commented:
Thanks!
